pkg/reporting/store.go

```diff
@@ -0,0 +1,168 @@
+	err := r.WithTransaction(ctx, func(tx *sqlx.Tx) error {
+		for _, product := range products {
+			_, err := EnsureProduct(ctx, tx, product)
+			if err != nil {
+				return fmt.Errorf("product ensure: %w", err)
+			}
+		}
+		for _, discount := range discounts {
+			_, err := EnsureDiscount(ctx, tx, discount)
+			if err != nil {
+				return fmt.Errorf("discount ensure: %w", err)
+			}
+		}
+		for _, query := range queries {
+			_, err := EnsureQuery(ctx, tx, query)
+			if err != nil {
+				return fmt.Errorf("query ensure: %w", err)
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("initialize: %w", err)
+	}
+	return nil
+}
+
+type Record struct {
+	TenantSource   string
+	CategorySource string
+	BillingDate    time.Time
+	ProductSource  string
+	DiscountSource string
+	QueryName      string
+	Value          float64
+}
+
+func (r *Store) WriteRecord(ctx context.Context, record Record) error {
+	return r.WithTransaction(ctx, func(tx *sqlx.Tx) error {
+		tenant, err := EnsureTenant(ctx, tx, &db.Tenant{Source: record.TenantSource})
+		if err != nil {
+			return fmt.Errorf("EnsureTenant(%q): %w", record.TenantSource, err)
+		}
+
+		category, err := EnsureCategory(ctx, tx, &db.Category{Source: record.CategorySource})
+		if err != nil {
+			return fmt.Errorf("EnsureCategory(%q): %w", record.CategorySource, err)
+		}
+
+		dateTime, err := EnsureDateTime(ctx, tx, NewDateTime(record.BillingDate))
+		if err != nil {
+			return fmt.Errorf("EnsureDateTime(%q): %w", record.BillingDate, err)
+		}
+
+		product, err := GetBestMatchingProduct(ctx, tx, record.ProductSource, record.BillingDate)
+		if err != nil {
+			return fmt.Errorf("GetBestMatchingProduct(%q, %q): %w", record.ProductSource, record.BillingDate, err)
+		}
+
+		discount, err := GetBestMatchingDiscount(ctx, tx, record.DiscountSource, record.BillingDate)
+		if err != nil {
+			return fmt.Errorf("GetBestMatchingDiscount(%q, %q): %w", record.DiscountSource, record.BillingDate, err)
+		}
+
+		query, err := GetQueryByName(ctx, tx, record.QueryName)
+		if err != nil {
+			return fmt.Errorf("GetQueryByName(%q): %w", record.QueryName, err)
+		}
+
+		fact := NewFact(dateTime, query, tenant, category, product, discount, record.Value)
+		if !isFactUpdatable(ctx, tx, fact, record.Value) {
+			return nil
+		}
+
+		_, err = EnsureFact(ctx, tx, fact)
+		if err != nil {
+			return fmt.Errorf("EnsureFact: %w", err)
+		}
+		return nil
+	})
+}
+
+// isFactUpdatable makes sure that only missing data or higher quantity values are saved in the billing database
+func isFactUpdatable(ctx context.Context, tx *sqlx.Tx, f *db.Fact, value float64) bool {
+	logger := ctrl.LoggerFrom(ctx)
+
+	fact, _ := GetByFact(ctx, tx, f)
+	if fact == nil || fact.Quantity < value {
+		return true
+	}
+	logger.Info(fmt.Sprintf("skipped saving, higher or equal number already recorded in DB "+
+		"for this hour: saved: \"%v\", new: \"%v\"", fact.Quantity, value))
+	return false
+}
```

Comment:
Nitpick: AFAIK this functionality only existed for DBaaS before, objectstorage updated the fact unconditionally. But I don't think we intentionally did that before, so this should be fine.

Reply:
yes that's right. Since the check seems to make sense in general, I added it there too. That's one of the only actual thing changing but I really doubt it has any negative impact.

Reply:
For consistency reasons I'd like to keep it - if you agree?